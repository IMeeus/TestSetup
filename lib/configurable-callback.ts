import { DeepPartial } from "./deep-partial";

export class ConfigurableCallback<TConfig, TResult> {
  private defaultConfig: TConfig;
  private callback: (config: TConfig) => TResult;

  constructor(defaultConfig: TConfig, callback: (config: TConfig) => TResult) {
    this.defaultConfig = defaultConfig;
    this.callback = callback;
  }

  run(override?: DeepPartial<TConfig> | ((config: TConfig) => void)): TResult {
    let config: TConfig;

    if (typeof override !== "function") {
      config = { ...this.defaultConfig, ...override };
    } else {
      config = { ...this.defaultConfig };
      override(config);
    }

    return this.callback(config);
  }
}
