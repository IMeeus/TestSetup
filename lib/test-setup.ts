import { DeepPartial } from "./deep-partial";

export class TestSetup<TConfig, TResult> {
  private defaultConfig: TConfig;
  private setupFunc: (config: TConfig) => TResult;

  constructor(defaultConfig: TConfig, setupFunc: (config: TConfig) => TResult) {
    this.defaultConfig = defaultConfig;
    this.setupFunc = setupFunc;
  }

  run(override?: DeepPartial<TConfig> | ((config: TConfig) => void)): TResult {
    let config = { ...this.defaultConfig };

    if (typeof override === "function") {
      override(this.defaultConfig);
    } else {
      config = { ...config, ...override };
    }

    return this.setupFunc(config);
  }
}
