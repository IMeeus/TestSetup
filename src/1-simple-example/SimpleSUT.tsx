export interface ISUTProps {
  title: string;
  items: string[];
}

export function SUT({ title, items }: ISUTProps) {
  return (
    <>
      <h1>{title}</h1>
      <ul>
        {items.map((i, index) => (
          <li key={index}>{i}</li>
        ))}
      </ul>
    </>
  );
}
