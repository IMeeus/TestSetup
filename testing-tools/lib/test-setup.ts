import { DeepPartial } from "./deep-partial";

export class TestSetup<TConfig, TResult> {
  private defaultConfig: TConfig;
  private setupFunc: (config: TConfig) => TResult;

  constructor(defaultConfig: TConfig, setupFunc: (config: TConfig) => TResult) {
    this.defaultConfig = defaultConfig;
    this.setupFunc = setupFunc;
  }

  run(overrides: DeepPartial<TConfig> = {}): TResult {
    return this.setupFunc({ ...this.defaultConfig, ...overrides });
  }

  run2(overrideFunc: (config: TConfig) => void = () => {}): TResult {
    const config = { ...this.defaultConfig };
    overrideFunc(config);

    return this.setupFunc(config);
  }
}
