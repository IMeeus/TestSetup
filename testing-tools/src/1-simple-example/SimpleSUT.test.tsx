import { render, screen } from "@testing-library/react";
import { SUT, ISUTProps } from "./SimpleSUT";
import { TestSetup } from "../../lib/test-setup";

const defaultProps: ISUTProps = {
  title: "someTitle",
  items: ["someItem1", "someItem2"],
};

const setup = new TestSetup(defaultProps, (props) => {
  return { ...render(<SUT {...props} />) };
});

test("renders correctly with default props", () => {
  const { container } = setup.run();

  expect(container).toMatchSnapshot();
});

test("overrides title correctly", () => {
  setup.run({
    title: "expectedTitle",
  });

  const title = screen.queryByText(/expectedTitle/i);
  expect(title).toBeInTheDocument();
});

test("overrides all items correctly", () => {
  setup.run({
    items: ["expectedItem"],
  });

  const expectedItem = screen.queryByText(/expectedItem/i);
  const someItem2 = screen.queryByText(/someItem2/i);

  expect(expectedItem).toBeInTheDocument();
  expect(someItem2).not.toBeInTheDocument();
});

test("overrides specific item correctly", () => {
  setup.run2((props) => {
    props.items[0] = "expectedItem";
  });

  const expectedItem = screen.queryByText(/expectedItem/i);
  const someItem2 = screen.queryByText(/someItem2/i);

  expect(expectedItem).toBeInTheDocument();
  expect(someItem2).toBeInTheDocument();
});
