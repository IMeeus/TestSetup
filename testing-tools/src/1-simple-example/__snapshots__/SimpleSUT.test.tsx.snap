// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with default props 1`] = `
<div>
  <h1>
    someTitle
  </h1>
  <ul>
    <li>
      someItem1
    </li>
    <li>
      someItem2
    </li>
  </ul>
</div>
`;
